File: CPP/Leet Code Solutions/valid-sudoku/valid-sudoku.cpp
bool isValidSudoku(vector<vector<char>>& board) 
{
		// creating 3 different hashset to maintain the check for rows, cols
		// and sub grid. We just need to check if row, col and subgrid 
		// doesnot have that char.
		
        unordered_set<char> row_visited[9];
        unordered_set<char> col_visited[9];
        unordered_set<char> subGrid_visited[3][3];
        
        for(int i=0;i<9;i++)
        {
            for(int j=0;j<9;j++)
            {
               char char_found = board[i][j];
			   
			   // if pos is blank then no issues, we had to check when it's not blank.
                if(char_found != '.') 
                {                     
                    if(row_visited[i].count(char_found) || col_visited[j].count(char_found) || subGrid_visited[i/3][j/3].count(char_found))
                        return false;
                }
                
				col_visited[j].insert(char_found);
				row_visited[i].insert(char_found);
				subGrid_visited[i / 3][j / 3].insert(char_found);
                
            }
        }
	 return true;
}

File: valid_sudoku.java
public boolean isValidSudoku(char[][] board) {
	if (board == null || board.length != 9 || board[0].length != 9)
		return false;
	for (int i = 0; i < 9; i++) {
		boolean[] m = new boolean[9];
		for (int j = 0; j < 9; j++) {
			if (board[i][j] != '.') {
				if (m[(int) (board[i][j] - '1')]) {
					return false;
				}
				m[(int) (board[i][j] - '1')] = true;
			}
		}
	}
 
	for (int j = 0; j < 9; j++) {
		boolean[] m = new boolean[9];
		for (int i = 0; i < 9; i++) {
			if (board[i][j] != '.') {
				if (m[(int) (board[i][j] - '1')]) {
					return false;
				}
				m[(int) (board[i][j] - '1')] = true;
			}
		}
	}
 
	for (int block = 0; block < 9; block++) {
		boolean[] m = new boolean[9];
		for (int i = block / 3 * 3; i < block / 3 * 3 + 3; i++) {
			for (int j = block % 3 * 3; j < block % 3 * 3 + 3; j++) {
				if (board[i][j] != '.') {
					if (m[(int) (board[i][j] - '1')]) {
						return false;
					}
					m[(int) (board[i][j] - '1')] = true;
				}
			}
		}
	}
 
	return true;
}
